src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -62,6 +64,12 @@ internal CommitManager(ImmutableArray<char> potentialCommitCharacters, RecentIte
                 && excludedCommitCharacter.Contains(typedChar));
         }
 
+        private ImmutableHashSet<char> GetPotentialCommitCharacters()
+            => _textView.Properties.TryGetProperty(CompletionSource.PotentialCommitCharacters, out ImmutableHashSet<char> potentialCommitCharacters)
+            ? potentialCommitCharacters
+            // If we were not initialized with a CompletionService or are called for a wrong textView, we should not make a commit.
+            : ImmutableHashSet<char>.Empty;
```

Comment:
needing to intersperse comments like this inside an expression is a pretty good sign the expression is too complex.

Reply:
i don't understand the ` If we were not initialized with a CompletionService` part.  How does it relate to this code?

Reply:
It's defensive coding. If some unrelated hackery or reflection goes bad, I guess we may get a different textview or something unusual like this.

Reply:
@cyrusnajmabadi, in the expected routine, CommitManager.PotentialCommitCharacters must be called after CompletionSource.InitializeCompletion set "PotentialCommitCharacters" property. However, we do not govern this. I would like to be protected if this assumption does not hold either in production or in test scenarios.

---
In reply to: [246444850](https://github.com/dotnet/roslyn/pull/32267#discussion_r246444850) [](ancestors = 246444850)